<a href="https://colab.research.google.com/github/revanthmadasu/machine-learning/blob/master/Lab_6_Template(1)_2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

   # Lab 6 - Template

The template is given as a starting point to help you get up to speed with loading file and understanding data.  
Use the template and answer each question in the appropriate cell.

Submit:
- eLearning

Naming: First-name_Last_name_ID_Lab_6.ipynb

## Dataset description

The dataset for Lab-6 is sample of the SBA dataset posted on Kaggle. 

The dataset is from the U.S. Small Business Administration (SBA)
The U.S. SBA was founded in 1953 on the principle of promoting and assisting small enterprises in the U.S. credit market 
(SBA Overview and History, US Small Business Administration (2015)). Small businesses have been a primary source of job creation in the United States; therefore, fostering small business formation and growth has social benefits by creating job opportunities and reducing unemployment.
There have been many success stories of start-ups receiving SBA loan guarantees such as FedEx and Apple Computer. 
However, there have also been stories of small businesses and/or start-ups that have defaulted on their SBA-guaranteed loans.

More info on the original dataset: https://www.kaggle.com/mirbektoktogaraev/should-this-loan-be-approved-or-denied


## Preparation

Load same data into H2O-3 dataframe.  
Describe imported dataframe.

In [2]:
! apt-get install default-jre
!java -version

Reading package lists... Done
Building dependency tree       
Reading state information... Done
default-jre is already the newest version (2:1.11-72).
default-jre set to manually installed.
0 upgraded, 0 newly installed, 0 to remove and 22 not upgraded.
openjdk version "11.0.18" 2023-01-17
OpenJDK Runtime Environment (build 11.0.18+10-post-Ubuntu-0ubuntu120.04.1)
OpenJDK 64-Bit Server VM (build 11.0.18+10-post-Ubuntu-0ubuntu120.04.1, mixed mode, sharing)


In [3]:
! pip install h2o

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 177.7/177.7 MB 7.2 MB/s eta 0:00:00


In [4]:
import h2o
try:
    h2o.cluster().shutdown()
except:
    pass 

In [5]:
# Adjust as per limits on your PC
# Limit to 8 threads and 8GB memory
h2o.init(nthreads=8, max_mem_size=8)

Checking whether there is an H2O instance running at http://localhost:54321..... not found.
Attempting to start a local H2O server...
  Java Version: openjdk version "11.0.18" 2023-01-17; OpenJDK Runtime Environment (build 11.0.18+10-post-Ubuntu-0ubuntu120.04.1); OpenJDK 64-Bit Server VM (build 11.0.18+10-post-Ubuntu-0ubuntu120.04.1, mixed mode, sharing)
  Starting server from /usr/local/lib/python3.9/dist-packages/h2o/backend/bin/h2o.jar
  Ice root: /tmp/tmpoex499rc
  JVM stdout: /tmp/tmpoex499rc/h2o_unknownUser_started_from_python.out
  JVM stderr: /tmp/tmpoex499rc/h2o_unknownUser_started_from_python.err
  Server is running at http://127.0.0.1:54321
Connecting to H2O server at http://127.0.0.1:54321 ... successful.


--------------------------  ----------------------------------
H2O_cluster_uptime:         06 secs
H2O_cluster_timezone:       Etc/UTC
H2O_data_parsing_timezone:  UTC
H2O_cluster_version:        3.40.0.1
H2O_cluster_version_age:    1 month
H2O_cluster_name:           H2O_from_python_unknownUser_eg18r8
H2O_cluster_total_nodes:    1
H2O_cluster_free_memory:    8 Gb
H2O_cluster_total_cores:    2
H2O_cluster_allowed_cores:  2
H2O_cluster_status:         locked, healthy
H2O_connection_url:         http://127.0.0.1:54321
H2O_connection_proxy:       {"http": null, "https": null}
H2O_internal_security:      False
Python_version:             3.9.16 final
--------------------------  ----------------------------------

In [8]:
df_h = h2o.import_file('./data/SBA_loans_lab_6.zip')

Parse progress: |████████████████████████████████████████████████████████████████| (done) 100%


In [9]:
print(df_h.columns)

['City', 'State', 'Zip', 'Bank', 'BankState', 'NAICS', 'Term', 'NoEmp', 'NewExist', 'CreateJob', 'RetainedJob', 'FranchiseCode', 'UrbanRural', 'RevLineCr', 'LowDoc', 'DisbursementGross', 'BalanceGross', 'GrAppv', 'SBA_Appv', 'Defaulted']


In [10]:
df_h.describe()

Rows:179833
Cols:20

City            State    Zip                Bank                            BankState    NAICS               Term                NoEmp               NewExist             CreateJob          RetainedJob         FranchiseCode       UrbanRural          RevLineCr    LowDoc    DisbursementGross    BalanceGross        GrAppv              SBA_Appv           Defaulted
-------  --------------  -------  -----------------  ------------------------------  -----------  ------------------  ------------------  ------------------  -------------------  -----------------  ------------------  ------------------  ------------------  -----------  --------  -------------------  ------------------  ------------------  -----------------  -------------------
type     enum            enum     int                enum                            enum         int                 int                 int                 int                  int                int                 int                 int                 enum         enum      int                  int                 int                 int                int
mins                              0.0                                                             0.0                 0.0                 0.0                 0.0                  0.0                0.0                 0.0                 0.0                                        0.0                  0.0                 500.0               250.0              0.0
mean                              53773.700544394                                                 399594.72817002475  110.90885988667281  11.384150851067377  1.2807185362326874   9.011527361496507  11.43381359372306   2755.5899473400377  0.7598827801349033                         201926.88862444597   6.553808255436987   193398.5314152572   150058.350658667   0.17522924046198418
maxs                              99999.0                                                         928120.0            461.0               8500.0              2.0                  8800.0             8800.0              99999.0             2.0                                        11446325.0           996262.0            5000000.0           4869000.0          1.0
sigma                             31182.08273423771                                               263331.7973921775   78.95806965689954   67.30104824719157   0.45206614960651126  246.6070749571537  247.46697982527462  12754.797773687804  0.6467730811038553                         290056.80003355315   2367.8805220862832  284469.67658896104  229388.8228276795  0.380164119046022
zeros                             43                                                              40267               161                 1369                220                  125596             87605               42030               64388                                      44                   179829              0                   0                  148321
missing  3               4        0                  316                             318          0                   0                   0                   23                   0                  0                   0                   0                   932          855       0                    0                   0                   0                  0
0        Cuyahoga Falls  OH       44224.0            PNC BANK, NATIONAL ASSOCIATION  OH           332710.0            84.0                3.0                 1.0                  0.0                3.0                 1.0                 1.0                 N            N         99000.0              0.0                 99000.0             49500.0            0.0
1        LAWRENCEVILLE   GA       30043.0            MERRILL LYNCH BANK USA          UT           454390.0            84.0                1.0                 2.0                  0.0                1.0                 1.0                 1.0                 0            N         100900.0             0.0 

### Set attributes for the categorical  variables - make non-numerical columns "factor" (categorical)

We will be using H2O-3 DRF model. The model can handle categorical variable and missing values.  
We will need to **mark** categorical variables with `asfactor` method

In [11]:
# Choose which columns to encode
cat_columns = ["City","State","Bank","BankState", "NewExist", "RevLineCr","LowDoc","Zip"]
encoded_columns = cat_columns
response = "Defaulted"

df_h[encoded_columns+[response]] = df_h[encoded_columns+[response]].asfactor()

In [12]:
df_h.describe()

Rows:179833
Cols:20

City            State    Zip    Bank                            BankState    NAICS               Term                NoEmp               NewExist    CreateJob          RetainedJob         FranchiseCode       UrbanRural          RevLineCr    LowDoc    DisbursementGross    BalanceGross        GrAppv              SBA_Appv           Defaulted
-------  --------------  -------  -----  ------------------------------  -----------  ------------------  ------------------  ------------------  ----------  -----------------  ------------------  ------------------  ------------------  -----------  --------  -------------------  ------------------  ------------------  -----------------  -----------
type     enum            enum     enum   enum                            enum         int                 int                 int                 enum        int                int                 int                 int                 enum         enum      int                  int                 int                 int                enum
mins                                                                                  0.0                 0.0                 0.0                             0.0                0.0                 0.0                 0.0                                        0.0                  0.0                 500.0               250.0
mean                                                                                  399594.72817002475  110.90885988667281  11.384150851067377              9.011527361496507  11.43381359372306   2755.5899473400377  0.7598827801349033                         201926.88862444597   6.553808255436987   193398.5314152572   150058.350658667
maxs                                                                                  928120.0            461.0               8500.0                          8800.0             8800.0              99999.0             2.0                                        11446325.0           996262.0            5000000.0           4869000.0
sigma                                                                                 263331.7973921775   78.95806965689954   67.30104824719157               246.6070749571537  247.46697982527462  12754.797773687804  0.6467730811038553                         290056.80003355315   2367.8805220862832  284469.67658896104  229388.8228276795
zeros                                                                                 40267               161                 1369                            125596             87605               42030               64388                                      44                   179829              0                   0
missing  3               4        0      316                             318          0                   0                   0                   23          0                  0                   0                   0                   932          855       0                    0                   0                   0                  0
0        Cuyahoga Falls  OH       44224  PNC BANK, NATIONAL ASSOCIATION  OH           332710.0            84.0                3.0                 1           0.0                3.0                 1.0                 1.0                 N            N         99000.0              0.0                 99000.0             49500.0            0
1        LAWRENCEVILLE   GA       30043  MERRILL LYNCH BANK USA          UT           454390.0            84.0                1.0                 2           0.0                1.0                 1.0                 1.0                 0            N         100900.0             0.0                 108400.0            81300.0            0
2        MAPLE GROVE     MN       55369  TWIN CITIES-METRO CERT. DEVEL   MN           0.0                 240.0               6.0                 1           4.0                6.0                 1.0                 0.0                 N            N         186000.0          

## Split to train/test/validation

In [13]:
train,test,valid = df_h.split_frame(ratios=[.7, .15], seed=123)

In [14]:
test

City             State      Zip  Bank                            BankState      NAICS    Term    NoEmp    NewExist    CreateJob    RetainedJob    FranchiseCode    UrbanRural  RevLineCr    LowDoc      DisbursementGross    BalanceGross          GrAppv        SBA_Appv    Defaulted
---------------  -------  -----  ------------------------------  -----------  -------  ------  -------  ----------  -----------  -------------  ---------------  ------------  -----------  --------  -------------------  --------------  --------------  --------------  -----------
CHICKASHA        OK       73018  JPMORGAN CHASE BANK NATL ASSOC  IL                 0      84       10           1            0              0                0             0  N            N              750000                       0  750000          607500                    0
ONAMIA           MN       56359  WOODLANDS NATIONAL BANK         MN            113310      48        1           1            2              1                0             2  N            N              130000                       0  130000          117000                    0
HUNTSVILLE       AL       35816  SOUTHERN DEVELOPMENT COUNCIL    AL            811192     240       15           2           15              0                1             0  N            N              507000                       0  507000          507000                    0
LOS ANGELES      CA       90003  FIRST BANK                      CA            327211     300       37           1            0              0                1             0  0            N              370000                       0  370000          277500                    0
ROHNERT PARK     CA       94928  BANK OF AMERICA CALIFORNIA N.A  CA                 0      69        5           2            0              0                1             0  N            N              100000                       0  100000           90000                    0
HINESVILLE       GA       31313  COASTAL AREA DISTRICT DEVEL AU  GA            721110     240       13           1           13             13                1             2  N            N                   1.075e+06               0       1.075e+06       1.075e+06            0
COLLEGE STATION  TX       77845  BBCN BANK                       CA            453998      84        1           1            3              1                1             1  0            N                5000                       0    5000            2500                    0
GAP              PA       17527  FULTON BANK NATL ASSOC          PA            722110      75        9           2            0              0            70130             2  0            Y              140000                       0  140000          119000                    0
HIALEAH          FL       33010  MERCANTIL COMMERCEBANK NATL AS  FL                 0     180        3           1            0              0                1             0  N            N               70000                       0   70000           63000                    0
MINNEAPOLIS      MN       55418  U.S. BANK NATIONAL ASSOCIATION  OH            722211      84        5           1            0              5            78760             1  Y            N               21302                       0   10000            5000                    0
[26854 rows x 20 columns]

In [15]:
train

City            State      Zip  Bank                            BankState      NAICS    Term    NoEmp    NewExist    CreateJob    RetainedJob    FranchiseCode    UrbanRural  RevLineCr    LowDoc      DisbursementGross    BalanceGross    GrAppv    SBA_Appv    Defaulted
--------------  -------  -----  ------------------------------  -----------  -------  ------  -------  ----------  -----------  -------------  ---------------  ------------  -----------  --------  -------------------  --------------  --------  ----------  -----------
Cuyahoga Falls  OH       44224  PNC BANK, NATIONAL ASSOCIATION  OH            332710      84        3           1            0              3                1             1  N            N                       99000               0     99000       49500            0
LAWRENCEVILLE   GA       30043  MERRILL LYNCH BANK USA          UT            454390      84        1           2            0              1                1             1  0            N                      100900               0    108400       81300            0
MAPLE GROVE     MN       55369  TWIN CITIES-METRO CERT. DEVEL   MN                 0     240        6           1            4              6                1             0  N            N                      186000               0    186000      186000            0
DIAMOND BAR     CA       91765  AURORA BANK FSB                 CO            446191     300        2           1            0              2                1             1  0            N                      280500               0    280500      210375            0
WILMINGTON      CA       90744  CITIBANK, N.A.                  SD                 0     298        2           1            0              0                1             0  N            N                      261000               0    261000      208800            0
LAFAYETTE       LA       70503  WELLS FARGO BANK NATL ASSOC     SD            441120      84       19           1            0             19                1             1  Y            N                       86483               0     30000       15000            0
CHICAGO         IL       60647  JPMORGAN CHASE BANK NATL ASSOC  IL            511210      84        1           1            0              1                0             1  Y            N                       44065               0     10000        5000            0
PARSONS         KS       67357  GREAT SOUTHERN BANK             KS            541921     120        3           1            0              0                1             0  N            N                       75000               0     75000       67500            0
ZANESVILLE      OH       43701  THE CITIZENS SAVINGS BANK       OH            235210      60        1           2            0              0                1             0  N            Y                       10000               0     10000        8000            0
DENVER          CO       80211  WELLS FARGO BANK NATL ASSOC     CO            443111     240       10           1            0              0                1             0  N            N                      170000               0    170000      127500            0
[125969 rows x 20 columns]

In [16]:
valid

City        State      Zip  Bank                            BankState      NAICS    Term    NoEmp    NewExist    CreateJob    RetainedJob    FranchiseCode    UrbanRural  RevLineCr    LowDoc      DisbursementGross    BalanceGross    GrAppv    SBA_Appv    Defaulted
----------  -------  -----  ------------------------------  -----------  -------  ------  -------  ----------  -----------  -------------  ---------------  ------------  -----------  --------  -------------------  --------------  --------  ----------  -----------
Gladstone   MO       64118  COMMERCE BANK                   MO            624120      63        3           2           20              3            38614             1  N            N                       40000               0     40000       20000            0
BEAUMONT    TX       77701  CAPITAL ONE NATL ASSOC          VA            454210      72        3           1            0              0                1             0  N            N                       60000               0     60000       48600            0
LANGHOME    PA       19047  BRANCH BK. & TR CO              NC            541940     120       20           1            5             20                1             1  0            N                      600000               0    600000      450000            0
CLINTON     OH       44216  JPMORGAN CHASE BANK NATL ASSOC  IL            488510      60       12           2            0             12                0             2  N            N                      141000               0    141000      126900            0
COVINA      CA       91722  COMERICA BANK                   TX            812112     300        5           1            2              5                1             1  0            N                      300600               0    300600      225450            0
FALLON      NV       89406  WELLS FARGO BANK NATL ASSOC     SD                 0      84       35           1            0             35                1             1  0            N                      224000               0    224000      168000            0
MIAMI       FL       33142  BANK OF AMERICA NATL ASSOC      NC            448150      84        3           1            0              0                1             1  Y            N                      100000               0    100000       50000            0
VENICE      FL       34285  CHARLOTTE STATE BANK & TRUST    FL                 0     300       13           2            0              0                1             0  N            N                      900000               0    900000      749700            0
KENT        WA       98032  WELLS FARGO BANK NATL ASSOC     SD            336612      87       22           1            0              0                1             1  N            N                      250000               0    250000      187500            0
LEWISVILLE  TX       75067  WELLS FARGO BANK NATL ASSOC     TX                 0     300       16           2            0              0            68060             0  N            N                      320000               0    330000      272000            0
[27010 rows x 20 columns]

## Question 1 

Use H2O-3 to train Distributed Random Forest model with default parameters.  
Documentation: https://docs.h2o.ai/h2o/latest-stable/h2o-docs/data-science/drf.html#defining-a-drf-model

The solution for the problem is provided.
You need to understand the command and the output in order to proceed to the next question.

In [ ]:
print(train.columns)

['City', 'State', 'Zip', 'Bank', 'BankState', 'NAICS', 'Term', 'NoEmp', 'NewExist', 'CreateJob', 'RetainedJob', 'FranchiseCode', 'UrbanRural', 'RevLineCr', 'LowDoc', 'DisbursementGross', 'BalanceGross', 'GrAppv', 'SBA_Appv', 'Defaulted']


In [ ]:
from h2o.estimators import H2ORandomForestEstimator

predictors = train.columns
predictors.remove("Defaulted")
print("Predictor columns:", predictors)    
response_col = "Defaulted"

drf = H2ORandomForestEstimator(nfolds=0,seed=123)

drf.train(predictors, response_col, training_frame= train)
model_summary = drf.summary()

Predictor columns: ['City', 'State', 'Zip', 'Bank', 'BankState', 'NAICS', 'Term', 'NoEmp', 'NewExist', 'CreateJob', 'RetainedJob', 'FranchiseCode', 'UrbanRural', 'RevLineCr', 'LowDoc', 'DisbursementGross', 'BalanceGross', 'GrAppv', 'SBA_Appv']
drf Model Build progress: |██████████████████████████████████████████████████████| (done) 100%


In [ ]:
model_summary

Model Summary: 
    number_of_trees    number_of_internal_trees    model_size_in_bytes    min_depth    max_depth    mean_depth    min_leaves    max_leaves    mean_leaves
--  -----------------  --------------------------  ---------------------  -----------  -----------  ------------  ------------  ------------  -------------
    50                 50                          5.29338e+06            20           20           20            5725          7517          6730.1

### Model performance metrics on Test dataset

Alternative method to calculate metrics is to use `h2o.make_metrics(pred, actual)` function:   
https://docs.h2o.ai/h2o/latest-stable/h2o-docs/performance-and-prediction.html#computing-model-metrics-from-general-predictions

In [ ]:
print("Best accuracy threshold:",drf.model_performance(test).accuracy()[0][0],"\n",
      " Accuracy:",
      drf.model_performance(test).accuracy()[0][1])
print("Best F1 threshold:",drf.model_performance(test).F1()[0][0],"\n",
      " F1:",
      drf.model_performance(test).F1()[0][1])
print("Model AUC:", drf.model_performance(test).auc())
print("Model AUCPR:", drf.model_performance(test).aucpr())
drf.model_performance(test).confusion_matrix()

Best accuracy threshold: 0.4525676262615756 
  Accuracy: 0.9347955611826916
Best F1 threshold: 0.34488910898743635 
  F1: 0.8054782794778516
Model AUC: 0.956193485901203
Model AUCPR: 0.8734735239591254


Confusion Matrix (Act/Pred) for max f1 @ threshold = 0.34488910898743635
       0      1     Error    Rate
-----  -----  ----  -------  ----------------
0      21272  871   0.0393   (871.0/22143.0)
1      947    3764  0.201    (947.0/4711.0)
Total  22219  4635  0.0677   (1818.0/26854.0)

In [ ]:
from h2o import make_metrics as H2OMakeMetrics
pred = drf.predict(test)[2]
print(pred)
actual = test["Defaulted"]
H2OMakeMetrics(pred, actual)


## Question 2

You are asked to tune the model from question #1. You should review full parameter list for the DRF model documentation: https://docs.h2o.ai/h2o/latest-stable/h2o-docs/data-science/drf.html#defining-a-drf-model

Pay special attention to the following parameters:
- balance_classes
- ntrees
- max_depth
- early stopping params: stopping_rounds, stopping_metric, stopping_tolerance
- mtries
- sample_rate

Alternative to search for the `ntrees` parameter is to set it to high number, and use early stopping to find optimal number of trees. (We will do it in the next question)

Find best parameters using H2O GridSearch for the following search space.
Modify the example code below as needed to make it work.  
You can experiment with additional options/parameters to see if you can get a better model.
```
# DRF hyperparameters
from h2o.grid.grid_search import H2OGridSearch

# DRF hyperparameters
drf_params2  = {'ntrees': [25,50,100],
               'max_depth': [10,15,20,25]}

# Train and validate a cartesian grid of DRFs
drf_grid2  = H2OGridSearch(model=H2ORandomForestEstimator,
                          grid_id='drf_grid2',
                          hyper_params=drf_params2)

drf_grid2.train(x=predictors,
               y=response_col,
               nfolds=0,
               training_frame=train,
               validation_frame=valid,
               seed=123)
```

You can follow the progress of the Grid search job by using H2O Flow, the URL is shown right after `h2o.init()`.
Answer following questions:
- What is the best set of parameters for the DRF model using AUCPR metric?
- What is the best model performance (AUCPR) on a validation dataset?
- What is the best model performance (AUCPR) on a test dataset?
- Produce Confusion Matrix using best F1 probability threshold

In [ ]:
# DRF hyperparameters
from h2o.grid.grid_search import H2OGridSearch

# DRF hyperparameters
drf_params2  = {'ntrees': [30,60],
                'sample_rate': [0.5, 0.7],
                'stopping_tolerance': [0, 0.003],
                'stopping_rounds': [0,3],
                'stopping_metric': ['logloss'],
                'mtries': [-1, -2], 
               'balance_classes': [True, False]}
# Train and validate a cartesian grid of DRFs
drf_grid2  = H2OGridSearch(model=H2ORandomForestEstimator,
                          grid_id='drf_grid2',
                          hyper_params=drf_params2)
drf_grid2.train(x=predictors,
               y=response_col,
               nfolds=0,
               training_frame=train,
               validation_frame=valid,
               seed=123)

drf Grid Build progress: |███████████████████████████████████████████████████████| (done) 100%


Hyper-Parameter Search Summary: ordered by increasing logloss
     balance_classes    mtries    ntrees    sample_rate    stopping_metric    stopping_rounds    stopping_tolerance    model_ids           logloss
---  -----------------  --------  --------  -------------  -----------------  -----------------  --------------------  ------------------  -------------------
     false              -1.0      60.0      0.7            logloss            0.0                0.0                   drf_grid2_model_14  0.18469905713867757
     false              -1.0      60.0      0.7            logloss            3.0                0.0                   drf_grid2_model_30  0.18469905713867757
     false              -1.0      60.0      0.7            logloss            0.0                0.003                 drf_grid2_model_46  0.18469905713867757
     false              -1.0      60.0      0.7            logloss            3.0                0.003                 drf_grid2_model_62  0.18469905713867757
     false              -1.0      60.0      0.5            logloss            3.0                0.0                   drf_grid2_model_22  0.18819536040603302
     false              -1.0      60.0      0.5            logloss            0.0                0.003                 drf_grid2_model_38  0.18819536040603302
     false              -1.0      60.0      0.5            logloss            3.0                0.003                 drf_grid2_model_54  0.18819536040603302
     false              -1.0      60.0      0.5            logloss            0.0                0.0                   drf_grid2_model_6   0.18819536040603302
     false              -1.0      30.0      0.7            logloss            0.0                0.0                   drf_grid2_model_10  0.18938217603931431
     false              -1.0      30.0      0.7            logloss            3.0                0.0                   drf_grid2_model_26  0.18938217603931431
---  ---                ---       ---       ---            ---                ---                ---                   ---                 ---
     false              -2.0      30.0      0.7            logloss            0.0                0.003                 drf_grid2_model_44  0.2731486035542063
     false              -2.0      30.0      0.7            logloss            3.0                0.003                 drf_grid2_model_60  0.2731486035542063
     true               -2.0      60.0      0.7            logloss            0.0                0.0                   drf_grid2_model_15  0.2782520361962055
     true               -2.0      60.0      0.7            logloss            3.0                0.0                   drf_grid2_model_31  0.2782520361962055
     true               -2.0      60.0      0.7            logloss            0.0                0.003                 drf_grid2_model_47  0.2782520361962055
     true               -2.0      60.0      0.7            logloss            3.0                0.003                 drf_grid2_model_63  0.2782520361962055
     true               -2.0      30.0      0.7            logloss            0.0                0.0                   drf_grid2_model_11  0.3005328075378542
     true               -2.0      30.0      0.7            logloss            3.0                0.0                   drf_grid2_model_27  0.3005328075378542
     true               -2.0      30.0      0.7            logloss            0.0                0.003                 drf_grid2_model_43  0.3005328075378542
     true               -2.0      30.0      0.7            logloss            3.0                0.003                 drf_grid2_model_59  0.3005328075378542
[64 rows x 10 columns]

In [ ]:
drf_gridperf1 = drf_grid2.get_grid(sort_by='aucpr', decreasing=True)
drf_gridperf1

Hyper-Parameter Search Summary: ordered by decreasing aucpr
     balance_classes    mtries    ntrees    sample_rate    stopping_metric    stopping_rounds    stopping_tolerance    model_ids           aucpr
---  -----------------  --------  --------  -------------  -----------------  -----------------  --------------------  ------------------  ------------------
     false              -2.0      60.0      0.5            logloss            3.0                0.0                   drf_grid2_model_24  0.8884518037530609
     false              -2.0      60.0      0.5            logloss            0.0                0.003                 drf_grid2_model_40  0.8884518037530609
     false              -2.0      60.0      0.5            logloss            3.0                0.003                 drf_grid2_model_56  0.8884518037530609
     false              -2.0      60.0      0.5            logloss            0.0                0.0                   drf_grid2_model_8   0.8884518037530609
     true               -2.0      60.0      0.5            logloss            3.0                0.0                   drf_grid2_model_23  0.882824657823606
     true               -2.0      60.0      0.5            logloss            0.0                0.003                 drf_grid2_model_39  0.882824657823606
     true               -2.0      60.0      0.5            logloss            3.0                0.003                 drf_grid2_model_55  0.882824657823606
     true               -2.0      60.0      0.5            logloss            0.0                0.0                   drf_grid2_model_7   0.882824657823606
     false              -2.0      60.0      0.7            logloss            0.0                0.0                   drf_grid2_model_16  0.881905365292187
     false              -2.0      60.0      0.7            logloss            3.0                0.0                   drf_grid2_model_32  0.881905365292187
---  ---                ---       ---       ---            ---                ---                ---                   ---                 ---
     true               -1.0      30.0      0.7            logloss            3.0                0.003                 drf_grid2_model_57  0.8736375522822927
     true               -1.0      30.0      0.7            logloss            0.0                0.0                   drf_grid2_model_9   0.8736375522822927
     true               -1.0      30.0      0.5            logloss            0.0                0.0                   drf_grid2_model_1   0.8732499758942533
     true               -1.0      30.0      0.5            logloss            3.0                0.0                   drf_grid2_model_17  0.8732499758942533
     true               -1.0      30.0      0.5            logloss            0.0                0.003                 drf_grid2_model_33  0.8732499758942533
     true               -1.0      30.0      0.5            logloss            3.0                0.003                 drf_grid2_model_49  0.8732499758942533
     false              -1.0      30.0      0.5            logloss            3.0                0.0                   drf_grid2_model_18  0.8714054108148194
     false              -1.0      30.0      0.5            logloss            0.0                0.0                   drf_grid2_model_2   0.8714054108148194
     false              -1.0      30.0      0.5            logloss            0.0                0.003                 drf_grid2_model_34  0.8714054108148194
     false              -1.0      30.0      0.5            logloss            3.0                0.003                 drf_grid2_model_50  0.8714054108148194
[64 rows x 10 columns]

### What is the best set of parameters for the DRF model using AUCPR metric?
DRF Model with paramenters balance_classes = false, mtries = -2, ntrees = 60, sample_rate = 0.5, stopping_rounds = (0,3), stopping_tolerance = (0.0, 0.003) performs best using aucpr metric

In [ ]:
# best model based on aucpr
best_drf_aucpr = drf_gridperf1.models[0]
print(best_drf_aucpr.summary())

# best model performance (AUCPR) on a validation dataset
print("Model AUCPR on validation set:", best_drf_aucpr.model_performance(valid).aucpr())

# best model performance (AUCPR) on a test dataset
print("Model AUCPR on test set:", best_drf_aucpr.model_performance(test).aucpr())

# confusion matrix
best_drf_aucpr.model_performance(test).confusion_matrix()

Model Summary: 
    number_of_trees    number_of_internal_trees    model_size_in_bytes    min_depth    max_depth    mean_depth    min_leaves    max_leaves    mean_leaves
--  -----------------  --------------------------  ---------------------  -----------  -----------  ------------  ------------  ------------  -------------
    60                 60                          3.59096e+06            20           20           20            3402          3735          3564.22
Model AUCPR on validation set: 0.8884518037530609
Model AUCPR on test set: 0.8889673864617536


Confusion Matrix (Act/Pred) for max f1 @ threshold = 0.43347253561095933
       0      1     Error    Rate
-----  -----  ----  -------  ----------------
0      21309  834   0.0377   (834.0/22143.0)
1      853    3858  0.1811   (853.0/4711.0)
Total  22162  4692  0.0628   (1687.0/26854.0)

### What is the best model performance (AUCPR) on a validation dataset?
AUCPR for best model on validation set is **0.8884518037530609**

### What is the best model performance (AUCPR) on a test dataset?

AUCPR for best model on test set is **0.8889673864617536**

### Produce Confusion Matrix using best F1 probability threshold
**Confusion Matrix (Act/Pred) for max f1 @ threshold = 0.43347253561095933**
0	1	Error	Rate
0	21309.0	834.0	0.0377	(834.0/22143.0)
1	853.0	3858.0	0.1811	(853.0/4711.0)
Total	22162.0	4692.0	0.0628	(1687.0/26854.0)


## Question 3

You are asked to tune the model from question #1. You should review full parameter list for the DRF model documentation: https://docs.h2o.ai/h2o/latest-stable/h2o-docs/data-science/drf.html#defining-a-drf-model

Tune model using early stopping criteria by specifying `ntrees`=1000 and setting the parameters required for early stopping. You will need to provide Validation data-frame for early stopping to work correctly.


Alternative to find optimal setting for `ntrees` (number of trees in RF) parameter is to set it to high number, and use early stopping to find optimal number of trees.

Find best parameters using H2O GridSearch for the following search space.
You should experiment with additional options/parameters to see if you can train better models.
The goal is to get a better model than in Q2. Add at least one additional hyper-parameter to tune.
Minimum parameters to tune:
```
drf_params  = {'max_depth': [10,15, 25]}
```
You should review early stopping parameters to be added to the `train` method:
- stopping_rounds
- stopping_metric
- stopping_tolerance

               
Answer following questions:
- What is the best set of parameters for the DRF model using AUCPR metric?
- What is the best model performance (AUCPR) on a validation dataset?
- What is the best model performance (AUCPR) on a test dataset?
- Produce Confusion Matrix using best F1 probability threshold

In [ ]:
# DRF hyperparameters
from h2o.grid.grid_search import H2OGridSearch

# DRF hyperparameters
drf_params3  = {'ntrees': [1000],
                'sample_rate': [0.5, 0.7, 0.9],
                'col_sample_rate_per_tree': [0.7, 0.8, 0.9],
                'mtries': [-2], 
                'balance_classes': [False],
                'max_depth': [10,15, 25]}
# Train and validate a cartesian grid of DRFs
drf_base = H2ORandomForestEstimator(seed=1234, nfolds=5, stopping_rounds=3, stopping_metric="AUCPR")
drf_grid3  = H2OGridSearch(model=drf_base,
                          grid_id='drf_grid2',
                          hyper_params=drf_params3)
drf_grid3.train(x=predictors,
               y=response_col,
               training_frame=train,
               validation_frame=valid,
               seed=123)

drf Grid Build progress: |███████████████████████████████████████████████████████| (done) 100%


Hyper-Parameter Search Summary: ordered by increasing logloss
     balance_classes    col_sample_rate_per_tree    max_depth    mtries    ntrees    sample_rate    model_ids           logloss
---  -----------------  --------------------------  -----------  --------  --------  -------------  ------------------  -------------------
     false              0.9                         15.0         -2.0      66.0      0.7            drf_grid2_model_15  0.16980973324255919
     false              0.9                         15.0         -2.0      60.0      0.9            drf_grid2_model_24  0.1710541782173759
     false              0.9                         15.0         -2.0      84.0      0.5            drf_grid2_model_6   0.17201149769489193
     false              0.9                         25.0         -2.0      73.0      0.7            drf_grid2_model_18  0.182090704764467
     false              0.9                         25.0         -2.0      79.0      0.5            drf_grid2_model_9   0.18220460528255955
     false              0.8                         15.0         -2.0      59.0      0.7            drf_grid2_model_14  0.18336699845874022
     false              0.9                         10.0         -2.0      56.0      0.7            drf_grid2_model_12  0.1835614386356362
     false              0.9                         10.0         -2.0      49.0      0.9            drf_grid2_model_21  0.18427061824967678
     false              0.9                         10.0         -2.0      68.0      0.5            drf_grid2_model_3   0.18437161707107763
     false              0.8                         15.0         -2.0      56.0      0.9            drf_grid2_model_23  0.18562943370990578
---  ---                ---                         ---          ---       ---       ---            ---                 ---
     false              0.7                         15.0         -2.0      56.0      0.9            drf_grid2_model_22  0.20207176820378103
     false              0.8                         10.0         -2.0      44.0      0.9            drf_grid2_model_20  0.20230999556666304
     false              0.8                         10.0         -2.0      49.0      0.7            drf_grid2_model_11  0.20282048804158057
     false              0.8                         10.0         -2.0      50.0      0.5            drf_grid2_model_2   0.20354566792960865
     false              0.7                         25.0         -2.0      69.0      0.7            drf_grid2_model_16  0.20428510111214984
     false              0.7                         15.0         -2.0      69.0      0.5            drf_grid2_model_4   0.20804403074755817
     false              0.7                         25.0         -2.0      74.0      0.5            drf_grid2_model_7   0.20805711934154208
     false              0.7                         10.0         -2.0      60.0      0.9            drf_grid2_model_19  0.2141330152470127
     false              0.7                         10.0         -2.0      65.0      0.7            drf_grid2_model_10  0.21470805816932537
     false              0.7                         10.0         -2.0      75.0      0.5            drf_grid2_model_1   0.21974269303303443
[27 rows x 9 columns]

In [ ]:
drf_gridperf3 = drf_grid3.get_grid(sort_by='aucpr', decreasing=True)
drf_gridperf3

Hyper-Parameter Search Summary: ordered by decreasing aucpr
     balance_classes    col_sample_rate_per_tree    max_depth    mtries    ntrees    sample_rate    model_ids           aucpr
---  -----------------  --------------------------  -----------  --------  --------  -------------  ------------------  ------------------
     false              0.8                         15.0         -2.0      59.0      0.7            drf_grid2_model_14  0.8893773797165693
     false              0.9                         25.0         -2.0      79.0      0.5            drf_grid2_model_9   0.8893548560455367
     false              0.8                         25.0         -2.0      61.0      0.7            drf_grid2_model_17  0.8889032627534529
     false              0.8                         25.0         -2.0      90.0      0.5            drf_grid2_model_8   0.888483059155926
     false              0.9                         15.0         -2.0      66.0      0.7            drf_grid2_model_15  0.8884652735721252
     false              0.9                         15.0         -2.0      84.0      0.5            drf_grid2_model_6   0.8883758470410299
     false              0.8                         15.0         -2.0      77.0      0.5            drf_grid2_model_5   0.8877279078309669
     false              0.8                         15.0         -2.0      56.0      0.9            drf_grid2_model_23  0.887273038932835
     false              0.9                         25.0         -2.0      73.0      0.7            drf_grid2_model_18  0.8870941786696642
     false              0.7                         15.0         -2.0      57.0      0.7            drf_grid2_model_13  0.8864340817556411
---  ---                ---                         ---          ---       ---       ---            ---                 ---
     false              0.9                         25.0         -2.0      51.0      0.9            drf_grid2_model_27  0.8825579156015583
     false              0.8                         10.0         -2.0      49.0      0.7            drf_grid2_model_11  0.8761924506192323
     false              0.8                         10.0         -2.0      44.0      0.9            drf_grid2_model_20  0.876149144449472
     false              0.7                         10.0         -2.0      65.0      0.7            drf_grid2_model_10  0.8752951808528368
     false              0.9                         10.0         -2.0      56.0      0.7            drf_grid2_model_12  0.875260365764241
     false              0.8                         10.0         -2.0      50.0      0.5            drf_grid2_model_2   0.8752520824328109
     false              0.7                         10.0         -2.0      60.0      0.9            drf_grid2_model_19  0.8749736795712676
     false              0.9                         10.0         -2.0      68.0      0.5            drf_grid2_model_3   0.8744743099227088
     false              0.9                         10.0         -2.0      49.0      0.9            drf_grid2_model_21  0.8744004909823139
     false              0.7                         10.0         -2.0      75.0      0.5            drf_grid2_model_1   0.8720940528479681
[27 rows x 9 columns]

### What is the best set of parameters for the DRF model using AUCPR metric?
DRF Model with paramenters balance_classes = false, mtries = -2, ntrees = 59, sample_rate = 0.5, stopping_rounds = 3, stopping_tolerance = 0.001 performs best using aucpr metric

In [ ]:
drf_gridperf3.models[0].summary()

Model Summary: 
    number_of_trees    number_of_internal_trees    model_size_in_bytes    min_depth    max_depth    mean_depth    min_leaves    max_leaves    mean_leaves
--  -----------------  --------------------------  ---------------------  -----------  -----------  ------------  ------------  ------------  -------------
    59                 59                          3.20323e+06            15           15           15            2792          4144          3281.36

In [ ]:
# best model based on aucpr
best_drf_aucpr = drf_gridperf3.models[0]
print(best_drf_aucpr.summary())

# best model performance (AUCPR) on a validation dataset
print("Model AUCPR on validation set:", best_drf_aucpr.model_performance(valid).aucpr())

# best model performance (AUCPR) on a test dataset
print("Model AUCPR on test set:", best_drf_aucpr.model_performance(test).aucpr())


Model Summary: 
    number_of_trees    number_of_internal_trees    model_size_in_bytes    min_depth    max_depth    mean_depth    min_leaves    max_leaves    mean_leaves
--  -----------------  --------------------------  ---------------------  -----------  -----------  ------------  ------------  ------------  -------------
    59                 59                          3.20323e+06            15           15           15            2792          4144          3281.36
Model AUCPR on validation set: 0.8960790581601458
Model AUCPR on test set: 0.8971408962438113


### What is the best model performance (AUCPR) on a validation dataset?
AUCPR for best model on validation set is **0.8960790581601458**

### What is the best model performance (AUCPR) on a test dataset?

AUCPR for best model on test set is **0.8971408962438113**

### Produce Confusion Matrix using best F1 probability threshold

In [ ]:
# confusion matrix
best_drf_aucpr.model_performance(test).confusion_matrix()

Confusion Matrix (Act/Pred) for max f1 @ threshold = 0.40259978218943787
       0      1     Error    Rate
-----  -----  ----  -------  ----------------
0      21458  685   0.0309   (685.0/22143.0)
1      861    3850  0.1828   (861.0/4711.0)
Total  22319  4535  0.0576   (1546.0/26854.0)

## Question 4

Use best model parameters from the GridSearch in question #3. Train new model as following:
- Train model on a combined dataset: train+validation
- Don't use cross-validation - `nfolds=0`
- Evaluate trained model on test dataset
- Compare AUCPR (test dataset) metrics from the question #2 and #3

**Training Model using combined dataset - Train + Validation without cross validation (nfolds=0)**

In [ ]:
combined = train.rbind(valid)
drf_4 = H2ORandomForestEstimator(nfolds=0,seed=123, balance_classes=False, mtries=-2, ntrees=59, sample_rate=0.7, col_sample_rate_per_tree=0.8, max_depth = 15)
drf_4.train(x=predictors,
            y=response_col,
            training_frame=combined)
model4_summary = drf_4.summary()

drf Model Build progress: |██████████████████████████████████████████████████████| (done) 100%


**Evaluate trained model on test dataset**

In [ ]:
print("Best accuracy threshold:",drf_4.model_performance(test).accuracy()[0][0],"\n",
      " Accuracy:",
      drf_4.model_performance(test).accuracy()[0][1])
print("Best F1 threshold:",drf_4.model_performance(test).F1()[0][0],"\n",
      " F1:",
      drf_4.model_performance(test).F1()[0][1])
print("Model AUC:", drf_4.model_performance(test).auc())
print("Model AUCPR:", drf_4.model_performance(test).aucpr())
drf_4.model_performance(test).confusion_matrix()

Best accuracy threshold: 0.4278081192121568 
  Accuracy: 0.9440679228420347
Best F1 threshold: 0.3900357060045771 
  F1: 0.8362541662186862
Model AUC: 0.9686558509764874
Model AUCPR: 0.8989451752728016


Confusion Matrix (Act/Pred) for max f1 @ threshold = 0.3900357060045771
       0      1     Error    Rate
-----  -----  ----  -------  ----------------
0      21442  701   0.0317   (701.0/22143.0)
1      822    3889  0.1745   (822.0/4711.0)
Total  22264  4590  0.0567   (1523.0/26854.0)

**Compare AURPC metrics from #2 and #3** 

AUCPR generated in this model is 0.898945 which is increased from 0.888968 from #2

AUCPR generated in this model is 0.898945 which is slightly increased from 0.897140 from #3


# Question 5

Review following documentation section first: https://docs.h2o.ai/h2o/latest-stable/h2o-docs/cross-validation.html

- Train DRF model with the parameters found in the Question 3.
- Use full data to train the model.
- Use cross-validation to produce reliable metrics for the model performance
- Produce model metrics using cross-validation 

**Note**: understand why the model training takes significantly longer than before we used cross-validation

**Training using full data and use cross validation to produce reliable metrics for model performance**

In [ ]:
all_data = train.rbind(valid).rbind(test)
drf_5 = H2ORandomForestEstimator(nfolds=7,
                                 seed=123,
                                 balance_classes=False,
                                 mtries=-2,
                                 ntrees=59,
                                 sample_rate=0.7,
                                 col_sample_rate_per_tree=0.8,
                                 max_depth = 15,
                                 keep_cross_validation_models = True)
trained_model = drf_5.train(x=predictors,
            y=response_col,
            training_frame=combined)
drf_5.summary()

drf Model Build progress: |██████████████████████████████████████████████████████| (done) 100%


Model Summary: 
    number_of_trees    number_of_internal_trees    model_size_in_bytes    min_depth    max_depth    mean_depth    min_leaves    max_leaves    mean_leaves
--  -----------------  --------------------------  ---------------------  -----------  -----------  ------------  ------------  ------------  -------------
    59                 59                          3.59907e+06            15           15           15            3059          4336          3657.71

**Produce model metrics using cross-validation**

In [ ]:
cv_metrics = drf_5.cross_validation_metrics_summary()
cv_metrics

Cross-Validation Metrics Summary: 
                         mean       sd           cv_1_valid    cv_2_valid    cv_3_valid    cv_4_valid    cv_5_valid    cv_6_valid    cv_7_valid
-----------------------  ---------  -----------  ------------  ------------  ------------  ------------  ------------  ------------  ------------
accuracy                 0.941874   0.00202452   0.941482      0.942623      0.943932      0.938859      0.94464       0.940267      0.941316
auc                      0.966868   0.00214012   0.967135      0.962598      0.968123      0.967683      0.966067      0.969366      0.967102
err                      0.0581259  0.00202452   0.0585184     0.057377      0.0560675     0.0611414     0.0553596     0.0597334     0.0586836
err_count                1270.43    49.2066      1286          1253          1229          1336          1193          1313          1283
f0point5                 0.83715    0.0106043    0.828202      0.846468      0.85496       0.826627      0.83857       0.828328      0.836896
f1                       0.832709   0.00410023   0.832726      0.828614      0.838438      0.827524      0.837266      0.830951      0.833441
f2                       0.828477   0.00899952   0.837301      0.811499      0.822543      0.828422      0.835966      0.833592      0.830015
lift_top_group           5.58345    0.127887     5.63837       5.66573       5.53253       5.44563       5.80965       5.48322       5.50899
logloss                  0.180172   0.00205108   0.179638      0.183142      0.178788      0.182009      0.176896      0.180224      0.180508
max_per_class_error      0.174234   0.0143109    0.159622      0.199524      0.187723      0.170978      0.164898      0.164639      0.172254
mcc                      0.797749   0.00524684   0.797316      0.794918      0.805173      0.790369      0.803918      0.794693      0.797855
mean_per_class_accuracy  0.896156   0.00532512   0.901529      0.886446      0.892467      0.895745      0.901131      0.898998      0.896776
mean_per_class_error     0.103844   0.00532512   0.0984712     0.113554      0.107533      0.104255      0.0988686     0.101003      0.103224
mse                      0.048599   0.000823657  0.0484672     0.0491052     0.0481846     0.0496219     0.0470865     0.0491386     0.0485888
pr_auc                   0.894083   0.00486925   0.894221      0.88806       0.900849      0.889045      0.895129      0.891838      0.899437
precision                0.840233   0.016541     0.825213      0.858804      0.866341      0.826031      0.839442      0.826588      0.839216
r2                       0.663624   0.00532987   0.66174       0.65721       0.672274      0.659244      0.66712       0.660779      0.667005
recall                   0.825766   0.0143109    0.840378      0.800476      0.812277      0.829022      0.835102      0.835361      0.827746
rmse                     0.220445   0.00187343   0.220153      0.221597      0.21951       0.22276       0.216994      0.221672      0.220429
specificity              0.966546   0.00446495   0.96268       0.972416      0.972658      0.962469      0.967161      0.962634      0.965805

**Note: It takes much longer time because cross validation for 7 folds creates 7+1 models and train them. So it take 7 times more time than without cross validation. And here we are training the model on full data which also takes additional time**